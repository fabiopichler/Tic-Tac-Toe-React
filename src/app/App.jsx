import React from 'react';

import Header from './header/Header';
import Board from './board/Board';
import Footer from './footer/Footer';

export const transpose = board => board[0].map((col, i) => board.map(row => row[i]));

export const checkRow = (row, player) => row.filter(col => col === player).length === 3;

export const checkRows = board => {
    for (const row of board)
        for (let player = 1; player <= 2; ++player)
            if (checkRow(row, player))
                return player;

    return 0;
}

export const checkDiagonals = board => {
    const diagonals = [[], []];

    for (const index in board) {
        diagonals[0].push(board[index][index]);
        diagonals[1].push(board[index][(board.length - 1) - index]);
    }

    return checkRows(diagonals);
}

export const check = (board, round) => {
    let player = checkRows(board);

    if (player === 0)
        player = checkRows(transpose(board));

    if (player === 0)
        player = checkDiagonals(board);

    if (player > 0)
        return player;

    if (player === 0 && round === 8)
        return -1;

    return 0;
}

const setupGame = () => ({
    player: 1,
    round: 0,
    winner: 0,
    board: [
        [0, 0, 0],
        [0, 0, 0],
        [0, 0, 0],
    ],
});

const App = () => {
    const [game, setGame] = React.useState(setupGame());

    const newGame = () => setGame(setupGame());

    const click = (col, rowIndex, colIndex) => {
        if (col !== 0 || game.winner !== 0)
            return;

        setGame(game => {
            const board = [...game.board];
            board[rowIndex][colIndex] = game.player;

            const winner = check(board, game.round);

            return {
                player: game.player === 1 ? 2 : 1,
                round: game.round + 1,
                winner,
                board,
            };
        });
    };

    return (
        <>
            <Header game={game} />

            <Board
                game={game}
                click={click}
            />

            <Footer
                game={game}
                onNewGame={newGame}
            />
        </>
    );
};

export default App;
