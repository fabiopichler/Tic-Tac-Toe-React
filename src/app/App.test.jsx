import React from 'react';
import renderer from 'react-test-renderer';

import App, {
    transpose,
    checkRow,
    checkRows,
    check,
} from './App';

test('transpose', () => {
    expect(transpose([
        [0, 0, 0],
        [1, 1, 1],
        [0, 0, 0],
    ])).toEqual([
        [0, 1, 0],
        [0, 1, 0],
        [0, 1, 0],
    ]);
});

test('checkRow', () => {
    expect(checkRow([1, 1, 1], 1)).toBe(true);
    expect(checkRow([1, 1, 0], 1)).toBe(false);
});

test('checkRows', () => {
    expect(checkRows([
        [0, 0, 0],
        [1, 1, 1],
        [0, 0, 0],
    ])).toEqual(1);
});

test('check', () => {
    expect(check([
        [0, 0, 0],
        [1, 1, 1],
        [0, 0, 0],
    ], 6)).toEqual(1);
});

test('renders correctly', () => {
    const component = renderer.create(
        <App />,
    );

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});
