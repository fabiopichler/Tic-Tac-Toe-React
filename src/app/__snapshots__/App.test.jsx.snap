// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App: renders correctly 1`] = `
Array [
  <header
    className="header my-3"
  >
    <h1
      className="text-center"
    >
      Tic Tac Toe
    </h1>
    <h5
      className="text-center mb-0"
    >
      Simples jogo da velha em React
    </h5>
    <div
      className="footer-players"
    >
      <div
        className="selected"
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-sun fa-w-16 mr-2"
          data-icon="sun"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M256 160c-52.9 0-96 43.1-96 96s43.1 96 96 96 96-43.1 96-96-43.1-96-96-96zm246.4 80.5l-94.7-47.3 33.5-100.4c4.5-13.6-8.4-26.5-21.9-21.9l-100.4 33.5-47.4-94.8c-6.4-12.8-24.6-12.8-31 0l-47.3 94.7L92.7 70.8c-13.6-4.5-26.5 8.4-21.9 21.9l33.5 100.4-94.7 47.4c-12.8 6.4-12.8 24.6 0 31l94.7 47.3-33.5 100.5c-4.5 13.6 8.4 26.5 21.9 21.9l100.4-33.5 47.3 94.7c6.4 12.8 24.6 12.8 31 0l47.3-94.7 100.4 33.5c13.6 4.5 26.5-8.4 21.9-21.9l-33.5-100.4 94.7-47.3c13-6.5 13-24.7.2-31.1zm-155.9 106c-49.9 49.9-131.1 49.9-181 0-49.9-49.9-49.9-131.1 0-181 49.9-49.9 131.1-49.9 181 0 49.9 49.9 49.9 131.1 0 181z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
        Player 1
      </div>
      <div
        className=""
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-moon fa-w-16 mr-2"
          data-icon="moon"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M283.211 512c78.962 0 151.079-35.925 198.857-94.792 7.068-8.708-.639-21.43-11.562-19.35-124.203 23.654-238.262-71.576-238.262-196.954 0-72.222 38.662-138.635 101.498-174.394 9.686-5.512 7.25-20.197-3.756-22.23A258.156 258.156 0 0 0 283.211 0c-141.309 0-256 114.511-256 256 0 141.309 114.511 256 256 256z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
        Player 2
      </div>
    </div>
  </header>,
  <main
    className="board"
  >
    <section
      className="board-container"
    >
      <div
        className="board-row row"
      >
        <div
          className="board-col col"
          onClick={[Function]}
        />
        <div
          className="board-col col"
          onClick={[Function]}
        />
        <div
          className="board-col col"
          onClick={[Function]}
        />
      </div>
      <div
        className="board-row row"
      >
        <div
          className="board-col col"
          onClick={[Function]}
        />
        <div
          className="board-col col"
          onClick={[Function]}
        />
        <div
          className="board-col col"
          onClick={[Function]}
        />
      </div>
      <div
        className="board-row row"
      >
        <div
          className="board-col col"
          onClick={[Function]}
        />
        <div
          className="board-col col"
          onClick={[Function]}
        />
        <div
          className="board-col col"
          onClick={[Function]}
        />
      </div>
    </section>
  </main>,
  <footer
    className="footer my-4"
  >
    <div
      className="text-center mb-3"
    >
      <button
        className="btn btn-primary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Reiniciar jogo
      </button>
    </div>
    <div
      className="text-center text-secondary small"
    >
      © 2020 Fábio Pichler
    </div>
  </footer>,
]
`;
